SpatialGDK/Source/SpatialGDK/Private/SpatialView/SubView.cpp

```diff
@@ -0,0 +1,217 @@
+﻿// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "SpatialView/SubView.h"
+
+#include "Utils/ComponentFactory.h"
+
+namespace SpatialGDK
+{
+SubView::SubView(const Worker_ComponentId TagComponentId, const FFilterPredicate Filter, const EntityView& View, FDispatcher& Dispatcher,
+				 const TArray<FDispatcherRefreshCallback> DispatcherRefreshCallbacks)
+	: TagComponentId(TagComponentId)
+	, Filter(Filter)
+	, View(View)
+	, TaggedEntities(TArray<Worker_EntityId>{})
+	, CompleteEntities(TArray<Worker_EntityId>{})
+	, NewlyCompleteEntities(TArray<Worker_EntityId>{})
+	, NewlyIncompleteEntities(TArray<Worker_EntityId>{})
+	, TemporarilyIncompleteEntities(TArray<Worker_EntityId>{})
+{
+	RegisterTagCallbacks(Dispatcher);
+	RegisterRefreshCallbacks(DispatcherRefreshCallbacks);
+}
+
+void SubView::TagQuery(Query& QueryToTag) const
+{
+	QueryToTag.ResultComponentIds.Add(TagComponentId);
+
+	QueryConstraint TagConstraint;
+	TagConstraint.ComponentConstraint = TagComponentId;
+
+	if (QueryToTag.Constraint.AndConstraint.Num() != 0)
+	{
+		QueryToTag.Constraint.AndConstraint.Add(TagConstraint);
+		return;
+	}
+
+	QueryConstraint NewConstraint;
+	NewConstraint.AndConstraint.Add(QueryToTag.Constraint);
+	NewConstraint.AndConstraint.Add(TagConstraint);
+
+	QueryToTag.Constraint = NewConstraint;
+}
+
+void SubView::TagEntity(TArray<FWorkerComponentData>& Components) const
+{
+	Components.Add(ComponentFactory::CreateEmptyComponentData(TagComponentId));
+}
+
+void SubView::Advance(const ViewDelta& Delta)
+{
+	// Note: Complete entities will be a longer list than the others for the majority of iterations under
+	// probable normal usage. This sort could then become expensive, and a potential optimisation would be
+	// to maintain the ordering of complete entities when merging in the newly complete entities and enforcing
+	// that complete entities is always sorted. This would also need to be enforced in the temporarily incomplete case.
+	// If this sort shows up in a profile it would be worth trying.
+	Algo::Sort(CompleteEntities);
+	Algo::Sort(NewlyCompleteEntities);
+	Algo::Sort(NewlyIncompleteEntities);
+	Algo::Sort(TemporarilyIncompleteEntities);
+
+	SubDelta.Project(Delta, CompleteEntities, NewlyCompleteEntities, NewlyIncompleteEntities, TemporarilyIncompleteEntities);
+
+	CompleteEntities.Append(NewlyCompleteEntities);
+	NewlyCompleteEntities.Empty();
+	NewlyIncompleteEntities.Empty();
+	TemporarilyIncompleteEntities.Empty();
+}
+
+const ViewDelta& SubView::GetViewDelta() const
+{
+	return SubDelta;
+}
+
+void SubView::RefreshEntity(const Worker_EntityId EntityId)
+{
+	if (TaggedEntities.Contains(EntityId))
+	{
+		CheckEntityAgainstFilter(EntityId);
+	}
+}
+
+FDispatcherRefreshCallback SubView::CreateComponentExistenceDispatcherRefreshCallback(
+	FDispatcher& Dispatcher, Worker_ComponentId ComponentId,
+	FComponentChangeRefreshPredicate RefreshPredicate = [](FEntityComponentChange) {
```

Comment:
I'm not convinced these predicates are useful without the current state of the completeness of the entity, at which point they become more of a third stage in the filtering pipeline, between the tag and filter.